### PR TITLE
[Feature][WIP] Support to write/read binlog

### PR DESCRIPTION
**Pull request by banmoy**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

File: be/src/storage/binlog_file_reader.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    Status seek(int64_t version, int64_t seq_id);
+
+    // Move to the next log entry.
+    // Return Status::Ok() if find the entry, Status::EndOfFile
+    // if there is no next, and other status if there is other error
+    Status next();
+
+    // Get the information of the current log entry
+    LogEntryInfo* log_entry();
+
```
Comment (LiShuMing) [inline be/src/storage/binlog_file_reader.h:74]: Is there an api to get  the consumed `seq_id`?

File: be/src/storage/binlog_file_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+#include "storage/binlog_file_writer.h"
+#include "storage/rowset/page_io.h"
+#include "util/coding.h"
+#include "util/crc32c.h"
+#include "util/raw_container.h"
+
+namespace starrocks {
+
+BinlogFileReader::BinlogFileReader(std::string file_name, std::shared_ptr<BinlogFileMetaPB> file_meta)
+        : _file_path(std::move(file_name)), _file_meta(file_meta) {}
```
Comment (LiShuMing) [inline be/src/storage/binlog_file_reader.cpp:27]: Why not `std::move(file_meta)`?  `file_meta` is shared across what?

File: be/src/storage/binlog_file_reader.cpp [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+}
+
+Status BinlogFileReader::_seek_to_page(int64_t version, int64_t seq_id) {
+    _reset_current_page_context();
+    do {
+        _reset_current_page_context();
+        PageHeaderPB& page_header_pb = _current_page_context->page_header;
+        RETURN_IF_ERROR(_read_page_header(_next_page_index, &page_header_pb));
+        if (page_header_pb.version() > version) {
+            return Status::NotFound(strings::Substitute("Can't find version $0", version));
```
Comment (LiShuMing) [inline be/src/storage/binlog_file_reader.cpp:77]: `fmt::format` is better? Add `page_header_pb.version()` in the error msg to debug?

File: be/src/storage/binlog_file_reader.cpp [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        }
+
+        // find the page containing the target change event
+        if (page_header_pb.version() == version && page_header_pb.end_seq_id() >= seq_id) {
+            RETURN_IF_ERROR(_read_page_content(_next_page_index, &page_header_pb, &(_current_page_context->page_content)));
+            break;
+        }
+
+        // reach the last page of the file
+        if (_next_page_index == _file_meta->num_pages()) {
```
Comment (LiShuMing) [inline be/src/storage/binlog_file_reader.cpp:87]: Is `if (_next_page_index + 1 == _file_meta->num_pages()) {` ?

File: be/src/storage/binlog_file_reader.cpp [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        return Status::InternalError("Fail to parse file header " + _file_path);
+    }
+    _next_page_index = 0;
+    _current_page_context = std::make_unique<PageContext>();
+    _current_log_entry = std::make_unique<LogEntryInfo>();
+
+    return _seek(version, seq_id);
+}
+
+Status BinlogFileReader::next() {
```
Comment (LiShuMing) [inline be/src/storage/binlog_file_reader.cpp:46]: Why not `Status next(LogEntryInfo* result_log_entry)` ?

File: be/src/storage/binlog_reader.cpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    LOG(INFO) << "Binlog reader " << _reader_id << ", seek to version " << version << ", seq_id " << seq_id;
+    return Status::OK();
+}
+
+Status BinlogReader::get_next(vectorized::ChunkPtr* chunk, int64_t max_version_exclusive) {
+    if (_next_version >= max_version_exclusive) {
+        return Status::EndOfFile("End of max version " + max_version_exclusive);
+    }
+
+    chunk->get()->reset();
```
Comment (LiShuMing) [inline be/src/storage/binlog_reader.cpp:60]: `chunk->get()->reset();` maybe not needed, `chunk` should be a new object?

File: be/src/storage/binlog_reader.cpp [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+    int seg_index = _log_entry_info->file_id->segment_index();
+    SegmentSharedPtr seg_ptr = _rowset->segments()[seg_index];
+    vectorized::SegmentReadOptions seg_options;
+    ASSIGN_OR_RETURN(seg_options.fs, FileSystem::CreateSharedFromString(_rowset->rowset_path()));
+    seg_options.chunk_size = _chunk_size;
+    vectorized::SparseRange range(start_row_id, seg_ptr->num_rows());
+    seg_options.rowid_range_option =
+            std::make_shared<vectorized::RowidRangeOption>(_rowset->rowset_id(), seg_index, range);
+    auto res = seg_ptr->new_iterator(_schema, seg_options);
```
Comment (LiShuMing) [inline be/src/storage/binlog_reader.cpp:155]: Can the same `_segment_iterator ` (the same segment file) be initialized multi time? maybe reuse _segment_iterator later?

Comment (LiShuMing) [reply be/src/storage/binlog_reader.cpp:155]: Does `new_iterator `  open other index files other than segment file?

File: be/src/storage/binlog_reader.cpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        _next_seq_id = 0;
+        return Status::OK();
+    }
+
+    Status status = _segment_iterator->get_next(chunk->get());
+    if (!status.ok() && !status.is_end_of_file()) {
+        return status;
+    }
+
+    // TODO append binlog meta columns
```
Comment (LiShuMing) [inline be/src/storage/binlog_reader.cpp:72]: How to add binlog meta?  Does BinlogReader support `filter`?

This maybe a problem, or `BinlogReader` just not support filter, always output full schema?

File: be/src/storage/binlog_manager.cpp [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+Status BinlogManager::init(Tablet& tablet) {
+    // TODO init binlog manager
+    // 1. restore binlog file meta
+    // 2. initialize metas about rowset
+    return Status::OK();
+}
+
+Status BinlogManager::add_insert_rowset(RowsetSharedPtr rowset) {
+    // TODO can block publish RPC
+    std::lock_guard lock(_write_lock);
```
Comment (mofeiatwork) [inline be/src/storage/binlog_manager.cpp:55]: such a big lock

File: be/src/storage/binlog_manager.cpp [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+            }
+            committed_writer_index = i;
+            last_writer = false;
+        }
+
+        // update file metas if all writers commit successfully
+        if (committed_writer_index == 0) {
+            std::vector<BinlogFileMetaPBSharedPtr> new_file_metas;
+            for (int i = 0; i < pending_writers.size(); i++) {
+                std::shared_ptr<BinlogFileWriter> writer = pending_writers[i];
```
Comment (mofeiatwork) [inline be/src/storage/binlog_manager.cpp:174]: Use reference to avoid copy `shared_ptr` ?

File: be/src/storage/binlog_manager.cpp [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+    }
+
+    if (status.ok() && rowset->num_rows() == 0) {
+        status = current_writer->add_empty();
+        if (!status.ok()) {
+            LOG(WARNING) << "Fail to add_empty for rowset " << rowset->rowset_id() << ", version "
+                         << rowset->start_version() << ", binlog writer " << current_writer->file_name() << ", "
+                         << status;
+        }
+        if (VLOG_IS_ON(3)) {
```
Comment (mofeiatwork) [inline be/src/storage/binlog_manager.cpp:146]: ?

File: be/src/storage/binlog_manager.cpp [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+
+    // abort the pending writer, and delete useless files
+    bool reuse_first_writer = _active_binlog_writer != nullptr;
+    std::vector<std::string> files_to_delete;
+    for (int i = 0; i < pending_writers.size(); i++) {
+        std::shared_ptr<BinlogFileWriter> writer = pending_writers[i];
+        if (i < committed_writer_index) {
+            Status st = writer->abort();
+            if (!st.ok()) {
+                LOG(WARNING) << "Fail to abort binlog writer " << writer->file_name();
```
Comment (mofeiatwork) [inline be/src/storage/binlog_manager.cpp:207]: how to handle this error ?

File: be/src/storage/binlog_manager.cpp [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+    return Status::InternalError(fmt::format("Fail to generate binlog for rowset {}, version {}",
+                                             rowset->rowset_id().to_string(), rowset->start_version()));
+}
+
+bool BinlogManager::is_rowset_used(const RowsetId& rowset_id) {
+    std::lock_guard lock(_meta_lock);
+    return _rowset_count_map.count(rowset_id) >= 1;
+}
+
+void BinlogManager::delete_expired_binlog() {
```
Comment (mofeiatwork) [inline be/src/storage/binlog_manager.cpp:240]: is there any daemon thread to do this work ?

File: be/src/storage/binlog_manager.cpp [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+                 << status;
+    Status st = binlog_writer->close();
+    if (!st.ok()) {
+        LOG(WARNING) << "Fail to close binlog writer, file id " << file_id << ", file name " << file_path << ", "
+                     << status;
+    }
+
+    std::shared_ptr<FileSystem> fs;
+    ASSIGN_OR_RETURN(fs, FileSystem::CreateSharedFromString(file_path))
+    st = fs->delete_file(file_path);
```
Comment (mofeiatwork) [inline be/src/storage/binlog_manager.cpp:289]: How to handle the delete failure

File: be/src/storage/binlog_file_writer.cpp [hunk 15]
```diff
@@ -0,0 +1,457 @@
...
+}
+
+Status BinlogFileWriter::add_insert_range(int32_t seg_index, int32_t start_row_id, int32_t num_rows) {
+    CHECK(_writer_state == PREPARE);
+    RETURN_IF_ERROR(_switch_page_if_full());
+    PendingPageContext* page_context = _pending_page_context.get();
+    LogEntryPB* log_entry = page_context->page_content.add_entries();
+    log_entry->set_entry_type(INSERT_RANGE_PB);
+    InsertRangePB* entry_data = log_entry->mutable_insert_range_data();
+    bool in_one_segment = false;
```
Comment (ABingHuang) [inline be/src/storage/binlog_file_writer.cpp:141]:
```suggestion
    bool in_one_segment = page_context->last_segment_index == seg_index;
```

File: be/src/storage/binlog_file_writer.cpp [hunk 16]
```diff
@@ -0,0 +1,457 @@
...
+    // if the last and current log entries are in the same segment, no need to set file id.
+    // When reading and iterating the page, we can get the file id from the last log entry
+    if (page_context->last_segment_index != seg_index) {
```
Comment (ABingHuang) [inline be/src/storage/binlog_file_writer.cpp:144]:
```suggestion
    if (!in_one_segment) {
```

File: be/src/storage/binlog_file_writer.h [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+    int seg_index;
+};
+
+// Binlog context for the current page to write
+struct PendingPageContext {
+    int64_t start_seq_id;
+    int64_t end_seq_id;
+    int32_t num_log_entries;
+    int32_t last_segment_index;
+    int32_t last_row_id;
```
Comment (ABingHuang) [inline be/src/storage/binlog_file_writer.h:35]: is it inclusive or exclusive?

File: be/src/storage/binlog_file_writer.h [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+
+struct RowsetSegInfo {
+    RowsetId& rowset_id;
+    int seg_index;
+};
+
+// Binlog context for the current page to write
+struct PendingPageContext {
+    int64_t start_seq_id;
+    int64_t end_seq_id;
```
Comment (ABingHuang) [inline be/src/storage/binlog_file_writer.h:32]: what are these first two fields used for?

File: gensrc/proto/binlog.proto [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+  // crc for compressed page
+  optional int32 compressed_page_crc = 5;
+  // the version of change events in the page
+  optional int64 version = 6;
+  // number of log entries
+  optional int32 num_log_entries = 7;
+  // sequence number of the first change event in the page
+  optional int64 start_seq_id = 8;
+  // sequence number of the last change event in the page
+  optional int64 end_seq_id = 9;
```
Comment (ABingHuang) [inline gensrc/proto/binlog.proto:91]: what are start_seq_id and end_seq_id used for?

File: be/src/storage/binlog_file_writer.h [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+    int32_t last_row_id;
+    int32_t estimated_page_size;
+    std::unordered_set<RowsetId, HashOfRowsetId> rowsets;
+    PageHeaderPB page_header;
+    PageContentPB page_content;
+};
+
+// Binlog context for the current version to write
+struct PendingVersionContext {
+    int64_t version;
```
Comment (ABingHuang) [inline be/src/storage/binlog_file_writer.h:44]: does version stand for epoch for binlog?

File: be/src/storage/binlog_file_writer.cpp [hunk 21]
```diff
@@ -0,0 +1,457 @@
...
+    file_meta->set_file_size(_file->size());
+    for (auto& rowset_id : version_context->rowsets) {
+        auto pair = _rowsets.emplace(rowset_id);
+        if (pair.second) {
+            RowsetIdPB* rowset_id_pb = file_meta->add_rowsets();
+            _set_row_id_pb(rowset_id, rowset_id_pb);
+        }
+    }
+    _reset_pending_context();
+    _writer_state = WAIT_WRITE;
```
Comment (ABingHuang) [inline be/src/storage/binlog_file_writer.cpp:261]: why state changed to WAIT_WRITE after commit?

**Review by ABingHuang — CHANGES_REQUESTED** [review]

File: be/src/storage/binlog_file_writer.cpp [hunk 22]
```diff
@@ -0,0 +1,457 @@
...
+    RETURN_IF_ERROR(_switch_page_if_full());
+    PendingPageContext* page_context = _pending_page_context.get();
+    LogEntryPB* log_entry = page_context->page_content.add_entries();
+    log_entry->set_entry_type(INSERT_RANGE_PB);
+    InsertRangePB* entry_data = log_entry->mutable_insert_range_data();
+    bool in_one_segment = false;
+    // if the last and current log entries are in the same segment, no need to set file id.
+    // When reading and iterating the page, we can get the file id from the last log entry
+    if (page_context->last_segment_index != seg_index) {
+        _set_file_id_pb(_pending_version_context->rowset_id, seg_index, entry_data->mutable_file_id());
```
Comment (ABingHuang) [inline be/src/storage/binlog_file_writer.cpp:145]: how to make sure the original data in rowset_id and seg_index still exist after compaction?
where to realize this logic?

File: be/src/storage/binlog_file_reader.cpp [hunk 23]
```diff
@@ -0,0 +1,348 @@
...
+    int64_t next_start_seq_id = log_entry_info->end_seq_id + 1;
+    int num_change_events = 0;
+    // assume next log entry share file id with the current first
+    FileIdPB* next_file_id = log_entry_info->file_id;
+    // assume rows in next log entry are continuous with current log entry first
+    int next_start_row_id = log_entry_info->start_row_id + log_entry_info->num_rows;
+    int num_rows = 0;
+    switch (next_log_entry->entry_type()) {
+    case INSERT_RANGE_PB: {
+        InsertRangePB* entry_data = next_log_entry->mutable_insert_range_data();
```
Comment (ABingHuang) [inline be/src/storage/binlog_file_reader.cpp:154]: mutable_xx is used to write. I think we need to read. should call next_log_entry->insert_range_data() here?

**Pull request closed without merging**